doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+
+## Version every trait
+Every trait gets its own disconnected version. This version would have
+to travel across the api-boundary along with the trait data. Whilst the obvious
+thing would be to enhance the low level `TraitsData` type to have a
+first class understanding of versioning, this cannot be so because not
+every method takes a `TraitsData` as an argument. The lowest-friction
+course of action might be to pack the version into `TraitId`, essentially
+doing string-munging to communicate the version across the api-boundary.
+
+Hosts and managers would need to be inspect the version data, which we
+could provide conveniences for, and construct the correct version
+of the trait view.
+
+```yml
+# traits.yml
+Camera:
+  description: >
+    The entity represents a device that captures a view through
+    re-projection or encoding into some other form.
+  usage:
+    - entity
+  version: 3
+```
+
+```python
+# Manager
+# In this example, CameraTrait is on version 3
+# SpatialTrait is on version 2
+#
+# The manager is responding to a resolve, needing to populate a
+# TraitsData with a camera trait and a spatial trait of the expected
+# version
+cameraVersion = mediacreation.versionFromTraitId(mediacreation.CameraTrait.kId)
+spatialVersion = mediacreation.versionFromTraitId(mediacreation.SpatialTrait.kId)
+
+# Support all known major versions.
+# Camera
+success_result = TraitsData()
+if (cameraVersion >= 3 and cameraVersion < 4):
+     trait = mediacreation.CameraTrait(success_result)
+     ... # Up to date camera population
+elif (cameraVersion >= 2 and cameraVersion < 1):
+     trait = mediacreation.legacy.CameraTrait_v2(success_result)
+     ... # V2 camera population
+elif (cameraVersion >= 1):
+     trait = mediacreation.legacy.CameraTrait_v1(success_result)
+     ... # V1 camera population
+else:
+  raise RuntimeError()
+
+# Spatial
+if (spatialVersion >= 2 and spatialVersion < 3):
+    trait = mediacreation.SpatialTrait(success_result)
+    ... # Up to date spatial population
+elif (spatialVersion >= 1):
+    trait = mediacreation.legacy.SpatialTrait_v1(success_result)
+     ... # V1 to date spatial population
+else:
+  raise RuntimeError()
+```
+
+### Benefits
+- Maximally flexible, hosts and managers can disregard versioning for
+  traits they don't care about.
+- Obvious from the YAML which version a trait is.
+- Individual traits will only update version when they change, thus they
+  will not have "gaps", they will always go v1->v2->v3.
+- A single trait can be updated in isolation without any effect on other
+  traits in the package.
+- Semver could be used to provide additional runtime compatibility
+  information.
+
+### Drawbacks
+- Version has to be encoded in `traitID` because several methods
+  (including `resolve`) take `traitSet`, which is only a collection of
+  `traitIDs`. There is no `TraitsData` to inspect. This would be a major
+  breaking change as currently conceived.
```

Comment:
Something else to consider - which I think might actually be quite important - is that we've often talked about "simple" managers that don't need first class understanding of every field in the data.  This probably extends to resolver caches too.

If the versioning information is _not_ part of the data, and is always driven by the host, then you end up effectively saying it its perhaps not even possible to write this kind manager any more. And that has generally been quite an important part of the API - that managers can store/recall host provided data without having to understand/manipulate this for each request. There may very well be only certain trait properties that are actively managed.

If the versioning is in the id, then you always know what version persisted data is/was, and in the a host is also capable of supporting older data as it _may be the only thing that is capable to mapping it to a newer schema version_.

Reply:
Some really good arguments for encoding the version in the trait ID there, I'm coming around to it, but still wary :thinking:

Reply:
What is the reticence?

Reply:
Downsides I can think of to encoding the version in the trait ID:

* Adding a property is a breaking change. I.e. version would have to be bumped, so then older trait views would think the trait is not imbued (at least as currently implemented).  If the version wasn't part of the trait ID, then views could detect if a trait is imbued regardless of version, and make a best effort to read/write properties they understand.  Of course, such a tolerant approach has its own downsides.
* Any trait with multiple versions would require branching around individually. If the schema version was agreed at initialisation, the host/manager has more opportunities to adapt their control flow ahead-of-time (e.g. effectively an entirely different plugin implementation could be constructed). Though both solutions _could_ live side-by-side.
* Specifications must be trait version aware, and these versions must be kept in sync somehow, or we require a rethink in how they're implemented and what they represent.

Reply:
> Putting it in the traits data instead of an ephemeral runtime quantity could be a middle ground, but then it precludes the use of any traitID centric introspection methods

Yeah we excluded `TraitsData` changes as a possibility for that reason.

> Would the expectation be that a manager has to then independently store and track the version in use at the time of storage - which kind of gets you back to what you get with it being in the id in the first place?

Well yes, when we chatted with Ynput about this exact question they said it's not ideal, but they could store the trait version alongside the trait ID.

I also assumed that it's the manager's responsibility to massage data from its backend, whether structured as traits or their own bespoke structure, into a form suitable for the host. But this assumption does have negative implications for the ideal of a "pass-through" manager scenario, as you brilliantly described elsewhere (i.e. https://github.com/OpenAssetIO/OpenAssetIO/pull/1254#discussion_r1495946276)

> You never see specs at the transport layer

This is a major wrinkle we wrestled with this option: how to ensure Specifications can be recognised across the API boundary, when there's no communication of Specification version, just the versions of the constituent traits.

If per-schema versioning was used, and that was encoded in the trait IDs, and all traits in a set use the same schema version, then you _could_ figure out the appropriate collection of possible Specification views.

If per-trait versioning was used, figuring out the appropriate Specification view is much harder. You would presumably have to iterate over all available MediaCreation packages, trying all their Specifications until one fit.

With the schema version communicated/agreed at initialisation time, you know immediately which Specifications are valid candidates.

Specifications need a rethink if trait IDs encode their version. If we remove the trait accessor methods (e.g. `.locatableContentTrait()`), then things become easier... E.g. Specifications could be redesigned to be trait version agnostic.